File: packages/oxlint-plugin-react-doctor/src/plugin/utils/visit-static-jsx-children.ts
import type { EsTreeNode } from "./es-tree-node.js";
import type { EsTreeNodeOfType } from "./es-tree-node-of-type.js";
import { isNodeOfType } from "./is-node-of-type.js";
import { stripParenExpression } from "./strip-paren-expression.js";

export interface StaticJsxChildrenCallbacks {
  // Called for every element reachable through static rendering structure.
  // Return false to skip the element's own children; anything else recurses.
  onElement: (element: EsTreeNodeOfType<"JSXElement">) => boolean | void;
  // Called for a child expression that renders content the walk cannot
  // statically enumerate ({children}, calls, member reads, spreads).
  onOpaqueExpression?: (expression: EsTreeNode) => void;
  // Called with the value of statically-known rendered text: non-whitespace
  // JSXText and string/number/template children.
  onStaticText?: (textValue: string) => void;
}

const visitExpression = (
  rawExpression: EsTreeNode,
  callbacks: StaticJsxChildrenCallbacks,
): void => {
  const expression = stripParenExpression(rawExpression);
  if (isNodeOfType(expression, "JSXElement")) {
    visitElement(expression, callbacks);
    return;
  }
  if (isNodeOfType(expression, "JSXFragment")) {
    visitStaticJsxChildren(expression.children, callbacks);
    return;
  }
  if (isNodeOfType(expression, "ConditionalExpression")) {
    visitExpression(expression.consequent, callbacks);
    visitExpression(expression.alternate, callbacks);
    return;
  }
  if (isNodeOfType(expression, "LogicalExpression")) {
    // `guard && <Jsx/>` renders the right side (the left renders only when
    // falsy, i.e. nothing-ish); `||` / `??` can render either side.
    if (expression.operator !== "&&") {
      visitExpression(expression.left, callbacks);
    }
    visitExpression(expression.right, callbacks);
    return;
  }
  if (isNodeOfType(expression, "ArrayExpression")) {
    for (const element of expression.elements) {
      if (!element) continue;
      if (isNodeOfType(element, "SpreadElement")) {
        callbacks.onOpaqueExpression?.(element);
      } else {
        visitExpression(element, callbacks);
      }
    }
    return;
  }
  if (isNodeOfType(expression, "Literal")) {
    if (typeof expression.value === "string" || typeof expression.value === "number") {
      const textValue = String(expression.value);
      if (textValue.trim().length > 0) callbacks.onStaticText?.(textValue);
    }
    return;
  }
  if (isNodeOfType(expression, "TemplateLiteral")) {
    // A template child renders text no matter what its expressions hold.
    const textValue = expression.quasis
      .map((quasi) => quasi.value.cooked ?? quasi.value.raw ?? "")
      .join(" ");
    if (textValue.trim().length > 0 || expression.expressions.length > 0) {
      callbacks.onStaticText?.(textValue);
    }
    return;
  }
  if (
    isNodeOfType(expression, "JSXEmptyExpression") ||
    (isNodeOfType(expression, "Identifier") && expression.name === "undefined")
  ) {
    return;
  }
  callbacks.onOpaqueExpression?.(expression);
};

const visitElement = (
  element: EsTreeNodeOfType<"JSXElement">,
  callbacks: StaticJsxChildrenCallbacks,
): void => {
  if (callbacks.onElement(element) === false) return;
  visitStaticJsxChildren(element.children, callbacks);
};

// Walks the statically-visible rendering structure below a JSX element's
// children: nested elements, fragments, and the JSX reachable through
// conditional, logical, and array expression children. Callbacks classify
// what the walker cannot: which elements match, which are opaque wrappers,
// and what rendered text means for the rule's claim.
export const visitStaticJsxChildren = (
  children: ReadonlyArray<EsTreeNode>,
  callbacks: StaticJsxChildrenCallbacks,
): void => {
  for (const child of children) {
    if (isNodeOfType(child, "JSXElement")) {
      visitElement(child, callbacks);
    } else if (isNodeOfType(child, "JSXFragment")) {
      visitStaticJsxChildren(child.children, callbacks);
    } else if (isNodeOfType(child, "JSXExpressionContainer")) {
      visitExpression(child.expression, callbacks);
    } else if (isNodeOfType(child, "JSXText")) {
      if (child.value.trim().length > 0) callbacks.onStaticText?.(child.value);
    }
  }
};

File: native/oxlint/rules/visit-static-jsx-children.rs
fn visit_static_jsx_children<'a, OnElement, OnOpaqueExpression, OnStaticText>(
    children: &'a [oxc_ast::ast::JSXChild<'a>],
    on_element: &mut OnElement,
    on_opaque_expression: &mut OnOpaqueExpression,
    on_static_text: &mut OnStaticText,
) where
    OnElement: FnMut(&'a oxc_ast::ast::JSXElement<'a>) -> bool,
    OnOpaqueExpression: FnMut(),
    OnStaticText: FnMut(),
{
    use oxc_ast::ast::JSXChild;

    for child in children {
        match child {
            JSXChild::Element(element) => {
                if on_element(element) {
                    visit_static_jsx_children(
                        &element.children,
                        on_element,
                        on_opaque_expression,
                        on_static_text,
                    );
                }
            }
            JSXChild::Fragment(fragment) => {
                visit_static_jsx_children(
                    &fragment.children,
                    on_element,
                    on_opaque_expression,
                    on_static_text,
                )
            }
            JSXChild::ExpressionContainer(container) => {
                if let Some(expression) = container.expression.as_expression() {
                    visit_static_jsx_expression(
                        expression,
                        on_element,
                        on_opaque_expression,
                        on_static_text,
                    );
                }
            }
            JSXChild::Text(text) if !text.value.trim().is_empty() => on_static_text(),
            JSXChild::Text(_) | JSXChild::Spread(_) => {}
        }
    }
}

fn visit_static_jsx_expression<'a, OnElement, OnOpaqueExpression, OnStaticText>(
    expression: &'a oxc_ast::ast::Expression<'a>,
    on_element: &mut OnElement,
    on_opaque_expression: &mut OnOpaqueExpression,
    on_static_text: &mut OnStaticText,
) where
    OnElement: FnMut(&'a oxc_ast::ast::JSXElement<'a>) -> bool,
    OnOpaqueExpression: FnMut(),
    OnStaticText: FnMut(),
{
    use oxc_ast::ast::{ArrayExpressionElement, Expression};
    use oxc_syntax::operator::LogicalOperator;

    let expression = expression.get_inner_expression();
    match expression {
        Expression::JSXElement(element) => {
            if on_element(element) {
                visit_static_jsx_children(
                    &element.children,
                    on_element,
                    on_opaque_expression,
                    on_static_text,
                );
            }
        }
        Expression::JSXFragment(fragment) => {
            visit_static_jsx_children(
                &fragment.children,
                on_element,
                on_opaque_expression,
                on_static_text,
            )
        }
        Expression::ConditionalExpression(conditional) => {
            visit_static_jsx_expression(
                &conditional.consequent,
                on_element,
                on_opaque_expression,
                on_static_text,
            );
            visit_static_jsx_expression(
                &conditional.alternate,
                on_element,
                on_opaque_expression,
                on_static_text,
            );
        }
        Expression::LogicalExpression(logical) => {
            if logical.operator != LogicalOperator::And {
                visit_static_jsx_expression(
                    &logical.left,
                    on_element,
                    on_opaque_expression,
                    on_static_text,
                );
            }
            visit_static_jsx_expression(
                &logical.right,
                on_element,
                on_opaque_expression,
                on_static_text,
            );
        }
        Expression::ArrayExpression(array) => {
            for element in &array.elements {
                if let Some(expression) = ArrayExpressionElement::as_expression(element) {
                    visit_static_jsx_expression(
                        expression,
                        on_element,
                        on_opaque_expression,
                        on_static_text,
                    );
                } else if matches!(element, ArrayExpressionElement::SpreadElement(_)) {
                    on_opaque_expression();
                }
            }
        }
        Expression::Identifier(identifier) if identifier.name == "undefined" => {}
        Expression::StringLiteral(string_literal) if !string_literal.value.trim().is_empty() => {
            on_static_text()
        }
        Expression::NumericLiteral(_) => on_static_text(),
        Expression::TemplateLiteral(template_literal)
            if !template_literal.expressions.is_empty()
                || template_literal.quasis.iter().any(|quasi| {
                    !quasi
                        .value
                        .cooked
                        .as_ref()
                        .map_or(quasi.value.raw.as_str(), |cooked| cooked.as_str())
                        .trim()
                        .is_empty()
                }) =>
        {
            on_static_text()
        }
        expression if expression.is_literal() => {}
        _ => on_opaque_expression(),
    }
}
